### PR TITLE
Clee/ma 77 auto course reruns

### PR DESCRIPTION
**Pull request by BenjiLee**
@nasthagiri Who else should I tag?

Going to combine the tickets. You can start the review on fad19b7.

### CODE REVIEW

**Comment by BenjiLee** [comment]
@nasthagiri b5d8b07 is also ready for review. I will post the related server side changes here.

File: edxval/api.py [hunk 1]
```diff
@@ -277,17 +278,63 @@ def get_videos_for_ids(
     return (VideoSerializer(video).data for video in videos)
 
 
-def get_video_info_for_course_and_profile(course_id, profile_name):
+def get_video_info_for_course_and_profiles(course_id, profiles):
     """
-    Returns a dict mapping profiles to URLs.
+    #TODO Change this to reflect new changes
```
Comment (nasthagiri) [inline edxval/api.py]: Is this TODO still needed?

File: edxval/api.py [hunk 2]
```diff
@@ -296,11 +343,47 @@ def get_video_info_for_course_and_profile(course_id, profile_name):
...
+    if Video.objects.filter(courses__course_id=unicode(new_course_id)):
+        raise ValCannotCreateError(
+            "New course id already exists: {}".format(new_course_id)
+        )
+    videos = Video.objects.filter(courses__course_id=unicode(old_course_id))
+
+    if len(videos) == 0:
+        raise ValVideoNotFoundError(
+            "Cannot find any videos for given course_id: {}".format(old_course_id)
+        )
```
Comment (nasthagiri) [inline edxval/api.py]: I don't think this error is needed.  We should support having courses without any videos.  We also don't want to require the caller to first check if the original course has gone through the video-pipeline process and has videos in VAL before calling this API.

File: edxval/tests/test_api.py [hunk 3]
```diff
@@ -436,3 +530,90 @@ def test_get_info_queries_for_only_video(self):
         with self.assertNumQueries(4):
             api.get_video_info(constants.VIDEO_DICT_ZEBRA.get("edx_video_id"))
 
+@ddt
+class RerunCourse(TestCase):
```
Comment (nasthagiri) [inline edxval/tests/test_api.py]: Rename the class to `TestRerunCourse`?

**Comment by BenjiLee** [comment]
@gwprice 

The other side of the world. https://github.com/edx/edx-platform/pull/7397

File: edxval/api.py [hunk 4]
```diff
@@ -277,17 +278,62 @@ def get_videos_for_ids(
...
-    Returns a dict mapping profiles to URLs.
+    Returns a dict of edx_video_ids with a dict of profiles.
 
     If the profiles or video is not found, urls will be blank.
+
+    Args:
+        course_id (str): id of the course
+        profiles (list): list of profile_names
+    Returns:
+        return_dict (dict): Returns all the profiles attached to a specific
```
Comment (gwprice) [inline edxval/api.py]: s/return_dict //

File: edxval/api.py [hunk 5]
```diff
@@ -277,17 +278,62 @@ def get_videos_for_ids(
...
+                },
+            },
+        }
+    Example:
+        Given two videos with two profiles each in course_id 'test_course':
+        >>> get_video_info_for_course_and_profiles('test-course', ['mobile', 'desktop''])
+        {
+            u'super-soaker': {
+                u'mobile': {
+                    'url': u'http: //www.meowmix.com',
```
Comment (gwprice) [inline edxval/api.py]: Use example.com for example URLs. Also, the example should distinguish between the URLs for the two videos. I suggest something like `http://example.com/super-soaker-mobile.mp4`

File: edxval/api.py [hunk 6]
```diff
@@ -296,11 +342,47 @@ def get_video_info_for_course_and_profile(course_id, profile_name):
...
+def rerun_course(old_course_id, new_course_id):
+    """
+    Adds the new_course_id to the videos taken from the old_course_id
+
+    Args:
+        old_course_id: The original course_id
+        new_course_id: The new course_id of the rerun
+    """
+    if old_course_id == new_course_id:
+        raise ValCannotCreateError(
```
Comment (gwprice) [inline edxval/api.py]: Just use `ValueError`, which is Python's general error for invalid arguments

File: edxval/api.py [hunk 7]
```diff
@@ -296,11 +342,47 @@ def get_video_info_for_course_and_profile(course_id, profile_name):
...
+                    "file_size": enc_vid.file_size,
+                    "duration": enc_vid.video.duration,
+                }
+            }
+        )
+
+    return return_dict
+
+
+def rerun_course(old_course_id, new_course_id):
```
Comment (gwprice) [inline edxval/api.py]: I think a more generic name like `copy_course_videos` would be better

File: edxval/api.py [hunk 8]
```diff
@@ -296,11 +342,47 @@ def get_video_info_for_course_and_profile(course_id, profile_name):
...
+
+    Args:
+        old_course_id: The original course_id
+        new_course_id: The new course_id of the rerun
+    """
+    if old_course_id == new_course_id:
+        raise ValCannotCreateError(
+            "Both course_id's are the same: {}".format(new_course_id)
+        )
+    if Video.objects.filter(courses__course_id=unicode(new_course_id)):
```
Comment (gwprice) [inline edxval/api.py]: I don't think this should be considered an error. The onus should be on the caller to be certain that they want to copy the videos, and I don't see a compelling reason to disallow this case.

File: edxval/api.py [hunk 9]
```diff
@@ -296,11 +342,47 @@ def get_video_info_for_course_and_profile(course_id, profile_name):
...
+        )
+    videos = Video.objects.filter(courses__course_id=unicode(old_course_id))
+
+    if len(videos) == 0:
+        raise ValVideoNotFoundError(
+            "Cannot find any videos for given course_id: {}".format(old_course_id)
+        )
+
+    for video in videos:
+        CourseVideo.objects.create(
```
Comment (gwprice) [inline edxval/api.py]: If the restriction against calling this function for a pre-existing course is removed, this should change to `get_or_create`

File: edxval/api.py (hunk 6 above)
Comment (gwprice) [inline edxval/api.py]: Alternatively, this could just be considered a no-op instead of an error

File: edxval/tests/test_api.py [hunk 10]
```diff
@@ -269,6 +275,94 @@ def test_get_url_for_profile(self):
...
+        """
+        Profile.objects.create(**constants.PROFILE_DICT_MOBILE)
+        Profile.objects.create(**constants.PROFILE_DICT_DESKTOP)
+        # 1st video
+        video = Video.objects.create(**constants.VIDEO_DICT_FISH)
+        EncodedVideo.objects.create(
+            video=Video.objects.get(
+                edx_video_id=constants.VIDEO_DICT_FISH.get("edx_video_id")
+            ),
+            profile=Profile.objects.get(profile_name="mobile"),
```
Comment (gwprice) [inline edxval/tests/test_api.py]: Save the result of `create` above to a variable and use it here instead of doing a get

File: edxval/tests/test_api.py [hunk 11]
```diff
@@ -269,6 +275,94 @@ def test_get_url_for_profile(self):
...
+
+        Creates two videos, each with two encodings.
+        """
+        Profile.objects.create(**constants.PROFILE_DICT_MOBILE)
+        Profile.objects.create(**constants.PROFILE_DICT_DESKTOP)
+        # 1st video
+        video = Video.objects.create(**constants.VIDEO_DICT_FISH)
+        EncodedVideo.objects.create(
+            video=Video.objects.get(
+                edx_video_id=constants.VIDEO_DICT_FISH.get("edx_video_id")
```
Comment (gwprice) [inline edxval/tests/test_api.py]: The "edx_video_id" key is guaranteed to be there, so use brackets instead of get

Comment (gwprice) [inline edxval/tests/test_api.py]: Never mind, this is about to be unnecessary

File: edxval/api.py (hunk 7 above)
Comment (BenjiLee) [inline edxval/api.py]: :) Was struggling to find a name for this.

File: edxval/tests/test_api.py [hunk 12]
```diff
@@ -269,6 +275,94 @@ def test_get_url_for_profile(self):
...
+        Creates a course for testing
+
+        Creates two videos, each with two encodings.
+        """
+        Profile.objects.create(**constants.PROFILE_DICT_MOBILE)
+        Profile.objects.create(**constants.PROFILE_DICT_DESKTOP)
+        # 1st video
+        video = Video.objects.create(**constants.VIDEO_DICT_FISH)
+        EncodedVideo.objects.create(
+            video=Video.objects.get(
```
Comment (gwprice) [inline edxval/tests/test_api.py]: Just use the video variable

File: edxval/api.py [hunk 13]
```diff
@@ -277,17 +278,62 @@ def get_videos_for_ids(
     return (VideoSerializer(video).data for video in videos)
 
 
-def get_video_info_for_course_and_profile(course_id, profile_name):
+def get_video_info_for_course_and_profiles(course_id, profiles):
     """
-    Returns a dict mapping profiles to URLs.
+    Returns a dict of edx_video_ids with a dict of profiles.
 
     If the profiles or video is not found, urls will be blank.
```
Comment (gwprice) [inline edxval/api.py]: This seems to be out of date. I think the description should explicitly state that a video without an encoding in one of the specified profiles will not be in the result.

File: edxval/tests/test_api.py [hunk 14]
```diff
@@ -269,6 +275,94 @@ def test_get_url_for_profile(self):
...
 
 
+class GetVideoForCourseProfiles(TestCase):
+    """Tests get_video_info_for_course_and_profiles in api.py"""
+
+    def setUp(self):
+        """
+        Creates a course for testing
+
+        Creates two videos, each with two encodings.
```
Comment (gwprice) [inline edxval/tests/test_api.py]: The test would be a bit stronger if if had a video in another course (to ensure that it gets filtered out)

File: edxval/tests/test_api.py [hunk 15]
```diff
@@ -269,6 +275,94 @@ def test_get_url_for_profile(self):
...
+    def test_get_video_for_course_profiles_repeated_profile(self):
+        """
+        Tests get_video_info_for_course_and_profiles with repeated profile
+        """
+        videos = api.get_video_info_for_course_and_profiles(
+            'test-course',
+            ['mobile', 'mobile'])
+        self.assertEqual(len(videos[constants.VIDEO_DICT_FISH.get("edx_video_id")]), 1)
+        self.assertEqual(len(videos[constants.VIDEO_DICT_STAR.get("edx_video_id")]), 1)
+
```
Comment (gwprice) [inline edxval/tests/test_api.py]: Too many blank lines here. Also, add test cases for specifying just one of the two profiles and for specifying an empty list of profiles

Comment (gwprice) [inline edxval/tests/test_api.py]: Ah, I now see that the single profile case is covered; I think it would make sense to split the _success method into multiple cases

File: edxval/tests/test_api.py [hunk 16]
```diff
@@ -269,6 +275,94 @@ def test_get_url_for_profile(self):
...
+        CourseVideo.objects.create(video=video, course_id=self.course_id)
+
+    def test_get_video_for_course_profiles_success(self):
+        """
+        Tests series of successful get_video_info_for_course_and_profiles calls
+        """
+        videos = api.get_video_info_for_course_and_profiles(
+            'test-course',
+            ['mobile'])
+        self.assertEqual(len(videos[constants.VIDEO_DICT_FISH.get("edx_video_id")]), 1)
```
Comment (gwprice) [inline edxval/tests/test_api.py]: There should be a stronger assertion here about the contents of the result.

```
self.assertEqual(
    videos,
    {
        video_dict["edx_video_id"]: {
            PROFILE_DICT_MOBILE["profile_name"]: {
                "url": ENCODED_VIDEO_DICT_MOBILE["url"],
                "file_size": ENCODED_VIDEO_DICT_MOBILE["file_size"],
                "duration": video_dict["duration"],
            }
        }
        for video_dict in [VIDEO_DICT_FISH, VIDEO_DICT_STAR]
    }
)
```

File: edxval/tests/test_api.py (hunk 14 above)
Comment (gwprice) [inline edxval/tests/test_api.py]: There should also be a video that has an encoding in one profile but not the other.

File: edxval/tests/test_api.py [hunk 17]
```diff
@@ -436,3 +530,90 @@ def test_get_info_queries_for_only_video(self):
...
 
+@ddt
+class TestRerunCourse(TestCase):
+    """Tests rerun course"""
+
+    def setUp(self):
+        """
+        Creates a course for testing
+
+        Creates two videos, each with two encodings.
```
Comment (gwprice) [inline edxval/tests/test_api.py]: EncodedVideos are unnecessary for this test

Comment (gwprice) [inline edxval/tests/test_api.py]: This should also create a video not in the test course to make sure that only videos in the test course get copied

File: edxval/tests/test_api.py [hunk 18]
```diff
@@ -436,3 +530,90 @@ def test_get_info_queries_for_only_video(self):
...
+        """
+        courses = VideoSerializer(video).data['courses']
+        self.assertEqual(sorted(courses), sorted(expected_course_list))
+
+    def test_successful_rerun(self):
+        """Tests a successful rerun and 2nd rerun"""
+        api.rerun_course('test-course', 'meow-rerun')
+        videos = Video.objects.all()
+        for video in videos:
+            self._check_video_course_list(video,
```
Comment (gwprice) [inline edxval/tests/test_api.py]: Fix whitespace here

File: edxval/api.py [hunk 19]
```diff
@@ -296,11 +341,41 @@ def get_video_info_for_course_and_profile(course_id, profile_name):
...
+        new_course_id: The new course_id of the rerun
+    """
+    if old_course_id == new_course_id:
+        raise ValueError(
+            "Both course_id's are the same: {}".format(new_course_id)
+        )
+    if Video.objects.filter(courses__course_id=unicode(new_course_id)):
+        raise ValCannotCreateError(
+            "New course id already exists: {}".format(new_course_id)
+        )
```
Comment (BenjiLee) [inline edxval/api.py]: I think there was a comment here that got deleted. @gwprice Did you say that we should allow both of the cases     `New course id already exists: {}".format(new_course_id)` and `"Both course_id's are the same: {}".format(new_course_id)`?

Comment (gwprice) [inline edxval/api.py]: Yes, I think we should allow the course ids to be the same and treat it as a no-op and allow target courses that already have videos. The onus should be on the caller to make sure that's what they really want.

File: edxval/api.py [hunk 20]
```diff
@@ -296,11 +341,41 @@ def get_video_info_for_course_and_profile(course_id, profile_name):
...
-        for enc_vid in encoded_videos
-    }
+    return_dict = {}
+    for enc_vid in encoded_videos:
+        return_dict.setdefault(enc_vid.video.edx_video_id, {}).update(
+            {
+                enc_vid.profile.profile_name: {
+                    "url": enc_vid.url,
+                    "file_size": enc_vid.file_size,
+                    "duration": enc_vid.video.duration,
```
Comment (gwprice) [inline edxval/api.py]: It's a little weird to include the duration with each encoding, when it's a property of the video and not the encoding. The fact that the sole intended caller of this ends up filtering out the duration suggests that we should consider modifying the returned object to something like the following:

```
{
    "video_id_1": {
        "duration": 100,
        "encodings": {
            "profile1": {
                "url": "http://example.com/foo.mp4",
                "file_size": 555,
            },
            ...
        }
    },
    ...
```

File: edxval/api.py [hunk 21]
```diff
@@ -296,11 +341,36 @@ def get_video_info_for_course_and_profile(course_id, profile_name):
...
+                    "url": enc_vid.url,
+                    "file_size": enc_vid.file_size,
+                }
+            }
+        )
+
+    return return_dict
+
+
+def copy_course_videos(old_course_id, new_course_id):
```
Comment (nasthagiri) [inline edxval/api.py]: Is this method idempotent?  If you call it again with the same argument values, it will throw an error, right?

File: edxval/tests/test_api.py [hunk 22]
```diff
@@ -436,3 +620,78 @@ def test_get_info_queries_for_only_video(self):
...
+        """Tests when the new course id name is the same as the original"""
+        # with self.assertRaises(ValueError):
+        #     api.copy_course_videos('test-course', 'test-course')
+        api.copy_course_videos('test-course', 'test-course')
+
+    def test_existing_new_course_id_for_rerun(self):
+        """Test when the new course id already exists"""
+        api.copy_course_videos('test-course', 'meow-rerun')
+        # with self.assertRaises(ValCannotCreateError):
+        #     api.copy_course_videos('test-course', 'meow-rerun')
```
Comment (nasthagiri) [inline edxval/tests/test_api.py]: should remove commented out code.

File: edxval/api.py [hunk 23]
```diff
@@ -296,11 +341,36 @@ def get_video_info_for_course_and_profile(course_id, profile_name):
...
+        old_course_id: The original course_id
+        new_course_id: The new course_id of the rerun
+    """
+    if old_course_id == new_course_id:
+        return
+
+    videos = Video.objects.filter(courses__course_id=unicode(old_course_id))
+
+    for video in videos:
+        CourseVideo.objects.create(
```
Comment (nasthagiri) [inline edxval/api.py]: this should be changed to `get_or_create` since you are allowing videos with the `new_course_id` to already exist.

File: edxval/api.py [hunk 24]
```diff
@@ -296,11 +341,36 @@ def get_video_info_for_course_and_profile(course_id, profile_name):
...
-            "duration": enc_vid.video.duration,
-        }
-        for enc_vid in encoded_videos
-    }
+    return_dict = {}
+    for enc_vid in encoded_videos:
+        return_dict.setdefault(enc_vid.video.edx_video_id, {}).update(
+            {
+                "duration": enc_vid.video.duration,
+                enc_vid.profile.profile_name: {
```
Comment (gwprice) [inline edxval/api.py]: This needs to be embedded in another dict to avoid conflicts between a profile name and "duration"

Comment (BenjiLee) [inline edxval/api.py]: I'm not sure I follow.

Comment (gwprice) [inline edxval/api.py]: If someone creates a profile with the name "duration", this will break. As per the original comment, I think it should be like the following:

```
{
    "video_id_1": {
        "duration": 100,
        "encodings": {
            "profile1": {
                "url": "http://example.com/foo.mp4",
                "file_size": 555,
            },
            ...
        }
    },
    ...
}
```

File: edxval/api.py (hunk 21 above)
Comment (gwprice) [inline edxval/api.py]: It is now idempotent.

Comment (nasthagiri) [inline edxval/api.py]: Yes, it is now.  When I had added that comment, the code was using `get` instead of `get_or_create`.  lgtm.

File: edxval/tests/test_api.py [hunk 25]
```diff
@@ -436,3 +622,74 @@ def test_get_info_queries_for_only_video(self):
         with self.assertNumQueries(4):
             api.get_video_info(constants.VIDEO_DICT_ZEBRA.get("edx_video_id"))
 
+
+class TestRerunCourse(TestCase):
```
Comment (nasthagiri) [inline edxval/tests/test_api.py]: Should be renamed to `TestCopyCourse`.

File: edxval/tests/test_serializers.py [hunk 26]
```diff
@@ -1,123 +1,123 @@
...
-        """
-        Tests an non dict input in the VideoSerializer
-        """
-        data = "hello"
-        serializer = VideoSerializer(data=data)
-        self.assertEqual(
-            serializer.errors.get("non_field_errors")[0],
-            "Invalid data"
-        )
+# # pylint: disable=E1101
```
Comment (nasthagiri) [inline edxval/tests/test_serializers.py]: Need to uncomment these tests.

**Comment by BenjiLee** [comment]
@nasthagiri @gwprice

File: edxval/api.py [hunk 27]
```diff
@@ -296,11 +339,39 @@ def get_video_info_for_course_and_profile(course_id, profile_name):
...
+    return return_dict
+
+
+def copy_course_videos(old_course_id, new_course_id):
+    """
+    Adds the new_course_id to the videos taken from the old_course_id
+
+    Args:
+        old_course_id: The original course_id
+        new_course_id: The new course_id of the rerun
```
Comment (nasthagiri) [inline edxval/api.py]: nitpick: I wouldn't say "rerun" here since this method is useful without it needing to be for the rerun feature.

File: edxval/tests/constants.py [hunk 28]
```diff
@@ -30,13 +30,33 @@
...
+)
+ENCODED_VIDEO_DICT_MOBILE3 = dict(
+    url="http://www.ssss.com",
+    file_size=1111,
+    bitrate=2222,
+)
+ENCODED_VIDEO_DICT_DESKTOP3 = dict(
+    url="http://www.hiss.com",
+    file_size=3333,
+    bitrate=4444,
```
Comment (nasthagiri) [inline edxval/tests/constants.py]: Nitpick: you don't necessarily need a  new encoded_video_dict for each encoding.  The same one can be reused, no? Otherwise, it adds extra baggage to the test code.

File: edxval/tests/test_api.py [hunk 29]
```diff
@@ -269,6 +270,191 @@ def test_get_url_for_profile(self):
...
+                }}
+            )
+        return video_dict
+
+    def test_get_video_for_course_profiles_success_one_profile(self):
+        """
+        Tests get_video_info_for_course_and_profiles for one profile
+        """
+        videos = api.get_video_info_for_course_and_profiles(
+            'test-course',
```
Comment (nasthagiri) [inline edxval/tests/test_api.py]: Nitpick: reuse `self.course_id` here.

File: edxval/tests/test_api.py [hunk 30]
```diff
@@ -436,3 +622,74 @@ def test_get_info_queries_for_only_video(self):
...
+        videos = Video.objects.all()
+        for video in videos:
+            self._check_video_course_list(video,
+                                          ['test-course',
+                                           'meow-rerun',
+                                           'woof-bark'])
+
+    def test_same_course_ids(self):
+        """Tests when the new course id name is the same as the original"""
+        api.copy_course_videos('test-course', 'test-course')
```
Comment (nasthagiri) [inline edxval/tests/test_api.py]: need to add assert here to make sure the list of courses for the video remains the same.

File: edxval/tests/test_api.py [hunk 31]
```diff
@@ -436,3 +622,74 @@ def test_get_info_queries_for_only_video(self):
...
+
+    def test_existing_new_course_id_for_rerun(self):
+        """Test when the new course id already exists"""
+        api.copy_course_videos('test-course', 'meow-rerun')
```
Comment (nasthagiri) [inline edxval/tests/test_api.py]: ?  where does the `meow-rerun` course-id already exist for the video?
Once fixed, need to add assert here to make sure the list of courses for the video remains the same.

File: edxval/api.py [hunk 32]
```diff
@@ -277,17 +278,59 @@ def get_videos_for_ids(
     return (VideoSerializer(video).data for video in videos)
 
 
-def get_video_info_for_course_and_profile(course_id, profile_name):
+def get_video_info_for_course_and_profiles(course_id, profiles):
     """
-    Returns a dict mapping profiles to URLs.
+    Returns a dict of edx_video_ids with a dict of requested profiles.
 
     If the profiles or video is not found, urls will be blank.
```
Comment (gwprice) [inline edxval/api.py]: This comment doesn't make sense; if a profile is missing, it will simply not be present. We should also note that a video with no encodings will not be included in the result.

File: edxval/api.py [hunk 33]
```diff
@@ -277,17 +278,59 @@ def get_videos_for_ids(
...
+    Args:
+        course_id (str): id of the course
+        profiles (list): list of profile_names
+    Returns:
+        (dict): Returns all the profiles attached to a specific
+        edx_video_id
+        {
+            edx_video_id: {
+                'duration': length of the video in seconds,
+                profile_name: {
```
Comment (gwprice) [inline edxval/api.py]: This needs to be updated (as does the example below)

File: edxval/tests/test_api.py [hunk 34]
```diff
@@ -269,6 +270,191 @@ def test_get_url_for_profile(self):
...
+        # 4th video different course
+        video = Video.objects.create(**constants.VIDEO_DICT_PLANT)
+        EncodedVideo.objects.create(
+            video=video,
+            profile=desktop_profile,
+            **constants.ENCODED_VIDEO_DICT_DESKTOP3
+        )
+        CourseVideo.objects.create(video=video, course_id=self.course_id2)
+
+    def _create_video_dict(self, video, profiles, encoded_videos):
```
Comment (gwprice) [inline edxval/tests/test_api.py]: Calls to this would be clearer if it took a dictionary from profile name to encoded video instead taking profiles and encoded_videos separately as lists.

File: edxval/tests/test_api.py (hunk 14 above)
Comment (gwprice) [inline edxval/tests/test_api.py]: Reminder: The latter two of these have not been addressed yet. (GitHub seems to have moved the first two from elsewhere.)

File: edxval/tests/test_api.py [hunk 35]
```diff
@@ -436,3 +622,74 @@ def test_get_info_queries_for_only_video(self):
...
+        self.assertEqual(sorted(courses), sorted(expected_course_list))
+
+    def test_successful_rerun(self):
+        """Tests a successful rerun and 2nd rerun"""
+        api.copy_course_videos('test-course', 'meow-rerun')
+        videos = Video.objects.all()
+        for video in videos:
+            self._check_video_course_list(video,
+                                          ['test-course',
+                                           'meow-rerun'])
```
Comment (gwprice) [inline edxval/tests/test_api.py]: Fix whitespace

File: edxval/tests/test_api.py [hunk 36]
```diff
@@ -436,3 +622,74 @@ def test_get_info_queries_for_only_video(self):
...
+
+        api.copy_course_videos('meow-rerun', 'woof-bark')
+        videos = Video.objects.all()
+        for video in videos:
+            self._check_video_course_list(video,
+                                          ['test-course',
+                                           'meow-rerun',
+                                           'woof-bark'])
```
Comment (gwprice) [inline edxval/tests/test_api.py]: Fix whitespace

File: edxval/tests/constants.py (hunk 28 above)
Comment (BenjiLee) [inline edxval/tests/constants.py]: Will address this when reorganizing constants/unit tests.

File: edxval/tests/test_api.py (hunk 34 above)
Comment (BenjiLee) [inline edxval/tests/test_api.py]: Will address this when reorganizing constants/unit tests.

File: edxval/tests/test_api.py (hunk 14 above)
Comment (gwprice) [inline edxval/tests/test_api.py]: This docstring needs to be updated

File: edxval/tests/test_api.py [hunk 37]
```diff
@@ -436,3 +641,75 @@ def test_get_info_queries_for_only_video(self):
...
+        api.copy_course_videos('test-course', 'course-rerun1')
+        original_videos = Video.objects.filter(courses__course_id='test-course')
+        copied_videos = Video.objects.filter(courses__course_id='course-rerun1')
+        other_course = Video.objects.filter(courses__course_id='test-course2')
+
+        self.assertEqual(len(original_videos), 2)
+        self.assertEqual(original_videos[0].edx_video_id,
+                         constants.VIDEO_DICT_FISH['edx_video_id'])
+        self.assertEqual(original_videos[1].edx_video_id,
+                         constants.VIDEO_DICT_STAR['edx_video_id'])
```
Comment (gwprice) [inline edxval/tests/test_api.py]: The order is not guaranteed. Use set comparison:

```
self.assertEqual(
    {original_video.edx_video_id for original_video in original_videos},
    {constants.VIDEO_DICT_FISH["edx_video_id"], constants.VIDEO_DICT_STAR["edx_video_id"]}
)
```

File: edxval/tests/test_api.py [hunk 38]
```diff
@@ -436,3 +641,75 @@ def test_get_info_queries_for_only_video(self):
         with self.assertNumQueries(4):
             api.get_video_info(constants.VIDEO_DICT_ZEBRA.get("edx_video_id"))
 
+
+class TestCopyCourse(TestCase):
+    """Tests rerun course"""
```
Comment (nasthagiri) [inline edxval/tests/test_api.py]: Consider using consistent terminology, replacing all "rerun"s with "copy"s.

File: edxval/tests/test_api.py [hunk 39]
```diff
@@ -436,3 +641,75 @@ def test_get_info_queries_for_only_video(self):
...
+        self.assertEqual(len(original_videos), 2)
+        self.assertEqual(original_videos[0].edx_video_id,
+                         constants.VIDEO_DICT_FISH['edx_video_id'])
+        self.assertEqual(original_videos[1].edx_video_id,
+                         constants.VIDEO_DICT_STAR['edx_video_id'])
+        self.assertTrue(set(original_videos) == set(copied_videos))
+
+        self.assertEqual(len(other_course), 1)
+        self.assertEqual(other_course[0].edx_video_id,
+                         constants.VIDEO_DICT_TREE['edx_video_id'])
```
Comment (gwprice) [inline edxval/tests/test_api.py]: I don't think this part is necessary; it's enough that the extra video does not appear in the copied_videos list.

File: edxval/tests/test_api.py [hunk 40]
```diff
@@ -436,3 +641,75 @@ def test_get_info_queries_for_only_video(self):
...
+    def test_same_course_ids(self):
+        """Tests when the new course id name is the same as the original"""
+        original_videos = Video.objects.filter(courses__course_id='test-course')
+        api.copy_course_videos('test-course', 'test-course')
+        copied_videos = Video.objects.filter(courses__course_id='test-course')
+        self.assertEqual(len(original_videos), 2)
+        self.assertEqual(original_videos[0].edx_video_id,
+                         constants.VIDEO_DICT_FISH['edx_video_id'])
+        self.assertEqual(original_videos[1].edx_video_id,
+                         constants.VIDEO_DICT_STAR['edx_video_id'])
```
Comment (gwprice) [inline edxval/tests/test_api.py]: I don't think these are really necessary; the only assertion that really matters is the one on the next line.

File: edxval/api.py [hunk 41]
```diff
@@ -296,11 +343,39 @@ def get_video_info_for_course_and_profile(course_id, profile_name):
...
+        return_dict[enc_vid.video.edx_video_id].setdefault("profiles", {}).update(
+            {enc_vid.profile.profile_name: {
+                "url": enc_vid.url,
+                "file_size": enc_vid.file_size,
+            }}
+        )
+    return return_dict
+
+
+def copy_course_videos(old_course_id, new_course_id):
```
Comment (gwprice) [inline edxval/api.py]: s/old/source/
s/new/dest/

File: edxval/tests/test_api.py [hunk 42]
```diff
@@ -436,3 +641,75 @@ def test_get_info_queries_for_only_video(self):
...
+        self.assertTrue(set(original_videos) == set(copied_videos))
+
+    def test_existing_new_course_id_for_rerun(self):
```
Comment (gwprice) [inline edxval/tests/test_api.py]: "existing_new_course_id" is kind of confusing, and we shouldn't refer to "rerun" here. I suggest test_existing_destination_course

File: edxval/tests/test_api.py [hunk 43]
```diff
@@ -436,3 +641,75 @@ def test_get_info_queries_for_only_video(self):
...
+        self.assertEqual(original_videos[1].edx_video_id,
+                         constants.VIDEO_DICT_STAR['edx_video_id'])
+        self.assertEqual(len(copied_videos), 3)
+
+        self.assertEqual(copied_videos[0].edx_video_id,
+                         constants.VIDEO_DICT_FISH['edx_video_id'])
+        self.assertEqual(copied_videos[1].edx_video_id,
+                         constants.VIDEO_DICT_STAR['edx_video_id'])
+        self.assertEqual(copied_videos[2].edx_video_id,
+                         constants.VIDEO_DICT_TREE['edx_video_id'])
```
Comment (gwprice) [inline edxval/tests/test_api.py]: Instead of these specific assertions, you can just do self.assertLessEqual(set(original_videos), set(copied_videos)); the <= operator on sets does subset comparison.

Comment (gwprice) [inline edxval/tests/test_api.py]: (I do think the length assertions are good to keep)

File: edxval/api.py [hunk 44]
```diff
@@ -296,11 +343,39 @@ def get_video_info_for_course_and_profile(course_id, profile_name):
...
+                "url": enc_vid.url,
+                "file_size": enc_vid.file_size,
+            }}
+        )
+    return return_dict
+
+
+def copy_course_videos(source_course_id, destination_course_id):
+    """
+    Adds the new_course_id to the videos taken from the old_course_id
```
Comment (gwprice) [inline edxval/api.py]: Update

File: edxval/api.py [hunk 45]
```diff
@@ -296,11 +343,39 @@ def get_video_info_for_course_and_profile(course_id, profile_name):
...
+
+    Args:
+        old_course_id: The original course_id
+        new_course_id: The new course_id where the videos will be copied
```
Comment (gwprice) [inline edxval/api.py]: Update

File: edxval/tests/test_api.py [hunk 46]
```diff
@@ -436,3 +641,57 @@ def test_get_info_queries_for_only_video(self):
         with self.assertNumQueries(4):
             api.get_video_info(constants.VIDEO_DICT_ZEBRA.get("edx_video_id"))
 
+
+class TestCopyCourse(TestCase):
+    """Tests copy_course_videos in api.py"""
+
+    def setUp(self):
+        """
+        Creates a a course with 2 videos and a course with 1 video
```
Comment (gwprice) [inline edxval/tests/test_api.py]: s/a a/a/

File: edxval/tests/test_api.py [hunk 47]
```diff
@@ -269,6 +270,210 @@ def test_get_url_for_profile(self):
...
+                constants.PROFILE_DICT_DESKTOP["profile_name"]: constants.ENCODED_VIDEO_DICT_DESKTOP2,
+            }
+        ))
+        self.assertEqual(videos, expected_dict)
+
+    def test_get_video_for_course_profiles_no_profile(self):
+        """Tests get_video_info_for_course_and_profiles with no such profile"""
+        videos = api.get_video_info_for_course_and_profiles(
+            'test-course',
+            ['no profile'])
```
Comment (nasthagiri) [inline edxval/tests/test_api.py]: do we allow spaces in profile names?  or did you mean this to be `no_profile`?

File: edxval/api.py [hunk 48]
```diff
@@ -277,17 +278,63 @@ def get_videos_for_ids(
...
+        }
+    Example:
+        Given two videos with two profiles each in course_id 'test_course':
+        {
+            u'edx_video_id_1': {
+                u'duration: 1111,
+                u'profiles': {
+                    u'mobile': {
+                    'url': u'http: //www.example.com/meow',
+                    'file_size': 2222
```
Comment (gwprice) [inline edxval/api.py]: Indent

**Comment by nasthagiri** [comment]
:+1: lgtm.  waiting to see cleanup of the test code via consolidation of constants.py in the future.

File: edxval/tests/test_api.py (hunk 47 above)
Comment (BenjiLee) [inline edxval/tests/test_api.py]: I don't think so. It should be no_profile.

**Comment by gwprice** [comment]
:+1:
